### ISSUE TITLE
diverging_sub_expression false negative on diverging body of block

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#10776]
### Summary

`diverging_sub_expression` triggers on `… || diverge()` but not `… || { diverge(); }`. It should trigger equivalently on both.

This came up due to https://github.com/rust-lang/rust/pull/104134 breaking this test in clippy:

https://github.com/rust-lang/rust-clippy/blob/a167973e8171bce22ded246283faaf80b729382d/tests/ui/diverging_sub_expression.stderr#L33-L37

due to the `panic!` expansion changing from `core::panicking::panic_fmt(…)` to `{ core::panicking::panic_fmt(…); }`.

### Lint Name

diverging_sub_expression

### Reproducer

**True positive:**

```rust
fn main() {
    let lhs = false;
    let _ = lhs || return;
}
```

```console
$ cargo clippy

warning: sub-expression diverges
 --> src/main.rs:3:20
  |
3 |     let _ = lhs || return;
  |                    ^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#diverging_sub_expression
  = note: `#[warn(clippy::diverging_sub_expression)]` on by default
```

---

**False negative:**

```rust
fn main() {
    let lhs = false;
    let _ = lhs || { return; };
}
```

```console
$ cargo clippy

... nothing
```

### Version

```text
rustc 1.71.0-nightly (2c41369ac 2023-05-13)
binary: rustc
commit-hash: 2c41369acc445d04129db40ba998dd7a89fb0d2e
commit-date: 2023-05-13
host: aarch64-apple-darwin
release: 1.71.0-nightly
LLVM version: 16.0.2
```
